### ISSUE TITLE
no-useless-escape false positive on chatacter classes including ]

### ISSUE THREAD

**Issue by YodaDaCoda** [eslint/eslint#11627]
**Tell us about your environment**

* **ESLint Version:** v5.16.0
* **Node Version:** v10.15.2
* **npm Version:** 5.8.0

**What parser (default, Babel-ESLint, etc.) are you using?** default

**Please show your full configuration:**

<details>
<summary>Configuration</summary>

```js
{
  "extends": [
    "eslint:recommended"
  ],
  "parserOptions": {
    "ecmaVersion": 2018
  },
  "env": {
    "browser": true
  },
  "globals": {
    "jQuery": "readonly"
  },
  "rules": {
    "no-global-assign": "error",
    "indent": ["warn", 2, {"MemberExpression": "off", "CallExpression": {"arguments": "off"}}],
    "linebreak-style": ["error", "unix"],
    "quotes": ["warn", "single"],
    "semi": ["error", "always"]
  }
}
```

</details>

**What did you do? Please include the actual source code causing the issue, as well as the command that you used to run ESLint.**

```js
var regex = (/[\[]/);
```

```bash
eslint deleteme.js
```

**What did you expect to happen?**
The regular expression should be permitted; the closing bracket `]` needs to be escaped within a character class, else it signifies the end of the character class.

**What actually happened? Please include the actual, raw output from ESLint.**
```
/mnt/data-ext/Work/Clouding Around/deleteme.js
  1:5   error  'regex' is assigned a value but never used  no-unused-vars
  1:16  error  Unnecessary escape character: \[            no-useless-escape

✖ 2 problems (2 errors, 0 warnings)
```

**Are you willing to submit a pull request to fix this bug?**
No

**Comment by platinumazure** [comment]
Hi @YodaDaCoda, thanks for the issue.

According to your example, the opening bracket is escaped, not the closing bracket. The opening bracket would not need to be escaped if it's already in a character class.

**Comment by YodaDaCoda** [comment]
My apologies, you are correct.

Upon further review, I realised that my IDE was highlighting well beyond the unnecessary escape sequence (because `eslint -f json` doesn't return an `endColumn` for the `no-unnecessary-escape` rule). In my attempt to produce a smaller test case for this report (my regex is ~150 chars), I'd bungled the issue.

Would it be possible to add `endColumn` to the result from `no-unnecessary-escape`? If so, should I make a separate issue for that, or change this issue?

Thanks for your assistance, and sorry again for the confusion.

**Comment by platinumazure** [comment]
@YodaDaCoda No worries. Either way is fine although I think a new issue might be slightly easier for the maintainers if it's not too much trouble. And yes, for something as precisely located as an unnecessary escape, I think we probably could/should include an end location. Thanks for clarifying and for informing us about this UX issue!

**Comment by YodaDaCoda** [comment]
Created #11629
